standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/properties/PropertyManager.java

```diff
@@ -0,0 +1,576 @@
+   * @param keys the key fragments
+   * @param value the new value or null if mapping should be removed
+   */
+  public void setProperty(String[] keys, Object value) {
+    // find schema from key (length)
+    PropertySchema schema = schemaOf(keys);
+    String mapKey = mapKey(keys);
+    PropertyMap dirtyMap;
+    final Map<String, PropertyMap> dirtyMaps = this.dirtyMaps;
+    synchronized (dirtyMaps) {
+      dirtyMap = dirtyMaps.get(mapKey);
+    }
+    PropertyMap map;
+    if (dirtyMap != null) {
+      map = dirtyMap;
+    } else {
+      // is is stored ?
+      map = store.fetchProperties(mapKey, s->schema);
+      if (map == null) {
+        // remove a value from a non persisted map, noop
+        if (value == null) {
+          return;
+        }
+        map = new PropertyMap(store, schema);
+      }
+    }
+    // map is not null
+    String propertyName = propertyName(keys);
+    if (value != null) {
+      map.putProperty(propertyName, value);
+    } else {
+      map.removeProperty(propertyName);
+    }
+    // if this is the first update to the map
+    if (map != dirtyMap) {
+      dirtyMaps.put(mapKey, map);
+    }
+  }
+
+  public Map<String, PropertyMap> selectProperties(String namePrefix, String predicateStr, String... projectStr) {
+    return selectProperties(namePrefix, predicateStr,
+        projectStr == null
+            ? Collections.emptyList()
+            : Arrays.asList(projectStr));
+  }
+
+  /**
+   * Selects a set of properties.
+   * @param mapPrefix the map name prefix
+   * @param selector the selector/transformer function
+   * @return the map of property maps keyed by their name
+   */
+  public Map<String, PropertyMap> selectProperties(String mapPrefix, Function<PropertyMap, PropertyMap> selector) {
+    final String mapKey = mapKey(mapPrefix);
+    final Map<String, PropertyMap> selected = store.selectProperties(mapKey,null, k->schemaOf(splitKey(k)) );
+    final Map<String, PropertyMap> maps = new TreeMap<>();
+    final Function<PropertyMap, PropertyMap> transform = selector == null? Function.identity() : selector;
+    selected.forEach((k, p) -> {
+      final PropertyMap dirtyMap = dirtyMaps.get(k);
+      final PropertyMap map = transform.apply(dirtyMap == null ? p.copy() : dirtyMap.copy());
+      if (map != null && !map.isEmpty()) {
+        maps.put(mapName(k), map);
+      }
+    });
+    return maps;
+  }
+  public Map<String, PropertyMap> selectProperties(String prefix, String predicateStr, List<String> projectStr) {
+    final JexlExpression predicate;
+    try {
+      predicate = JEXL.createExpression(predicateStr);
+    } catch(JexlException.Parsing xparse) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(predicateStr, xparse);
+      }
+      throw xparse;
+    }
+    final Function<PropertyMap, PropertyMap> transform = (map)->{
+      MapWrapper wrapped = new MapWrapper(map);
+      Object result;
+      try {
+        result = predicate.evaluate(wrapped);
```

Comment:
Not sure if it will bring potential security problem provided that the `predicateStr` can be arbitrary.

Reply:
No issue there, JEXL is configured in a tight manner (no side-effects, accessible classes limited to the restricted).